### ISSUE TITLE
Spurious E0308 in latest release and pre-release versions

### ISSUE THREAD

**Issue by ohell** [rust-lang/rust-analyzer#23344]
After latest system update on Arch, rust lang and VScode and extensions all got updated. Now there are many false E0308 errors. I am attaching the two simplest  cases below. I have to revert to version 0.3.2719 in order to silence this noise.

**rust-analyzer version**: 0.4.3045, 0.3.3041 both show these errors, 0.3.2719 does not

**rustc version**: 1.88.0 (6b00bc388 2025-06-23)

**editor or extension**: VSCode extension version 0.3041

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

**repository link (if public, optional)**: (eg. [rust-analyzer](https://github.com/rust-lang/rust-analyzer))

**code snippet to reproduce**:
```rust
impl std::fmt::Display for Counter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.get())    // E0308: expected Result<(), Error>, found ()
    }
}

pub fn open_reader(path: Option<&Path>) -> io::Result<Box<dyn BufRead + Send>> {
    // Treat None or "-" as stdin
    if path.is_none_or(|p| p.as_os_str() == "-") {
        return Ok(Box::new(BufReader::with_capacity(1 << 26, std::io::stdin())));   // E0308: expected Box<dyn BufRead + Send, Global>, found Box<BufReader<Stdin>, Global>
    }
/*
...
*/
}

```

**Comment by ChayimFriedman2** [comment]
Your Rust version is unsupported (the earliest supported version is 1.94.0, we should also show a warning), so while I don't know the cause of this, if it does not reproduce on later toolchains I will close this.
